app/src/main/java/protect/card_locker/importexport/CatimaImporter.java

```diff
@@ -189,8 +258,39 @@ public void parseV2Groups(SQLiteDatabase database, String data) throws IOExcepti
         }
 
         for (CSVRecord record : records) {
-            importGroup(database, record);
+            importGroupV2(database, record);
+        }
+    }
+
+    public Hashtable<Integer,String> parseV3Groups(SQLiteDatabase database, String data) throws IOException, FormatException, InterruptedException {
+        // Parse groups
+        final CSVParser groupParser = new CSVParser(new StringReader(data), CSVFormat.RFC4180.builder().setHeader().build());
+
+        List<CSVRecord> records = new ArrayList<>();
+
+        try {
+            for (CSVRecord record : groupParser) {
+                records.add(record);
+
+                if (Thread.currentThread().isInterrupted()) {
+                    throw new InterruptedException();
+                }
+            }
+        } catch (IllegalArgumentException | IllegalStateException e) {
+            throw new FormatException("Issue parsing CSV data", e);
+        } finally {
+            groupParser.close();
+        }
+
+        Hashtable<Integer,String> groupsTable = new Hashtable<Integer,String>();
+
+        for (CSVRecord record : records) {
+            Pair<Integer, String> group = importGroupV3(database, record);
+            Integer id = group.first;
+            String name = group.second;
+            groupsTable.put(id, name);
         }
+        return groupsTable;
```

Comment:
Why all of this? Why not just a simple `importGroupV3` function call here?

(Also, just now seeing how often I copy-pasted the whole reading CSV logic. It's repeated in card, group and cardgroup stuff. Ugh, past me why did you write ugly copy-paste code :sob:)

Reply:
It's better now :)